File: super_ttt/engine.py
"""超级井字棋（Ultimate Tic-Tac-Toe）核心规则引擎。

纯逻辑实现，不依赖任何 GUI 库，可直接单元测试。

数据模型
--------
cells[sub][cell] : 小格内 9 个位置。0=空, 1=圈, 2=叉
grids[sub]       : 大格状态。0=未决出, 1=圈占领, 2=叉占领, 3=平
forced           : int|None。下一手被强制落子的大格；None 表示可在所有未决出大格自由落子
turn             : 当前行动方。1=圈, 2=叉
winner           : 0=进行中, 1=圈胜, 2=叉胜, 3=平局
win_line         : 大棋盘三连的三个大格下标（获胜时非 None）
"""

from __future__ import annotations

EMPTY, CIRCLE, CROSS = 0, 1, 2
GRID_OPEN, GRID_CIRCLE, GRID_CROSS, GRID_TIE = 0, 1, 2, 3

WIN_LINES = (
    (0, 1, 2), (3, 4, 5), (6, 7, 8),
    (0, 3, 6), (1, 4, 7), (2, 5, 8),
    (0, 4, 8), (2, 4, 6),
)


def line_winner(states: list) -> tuple:
    """states: 长度为 9 的序列（格值 0/1/2 或大格值 0/1/2/3）。
    只有圈(1)/叉(2)能构成三连，平局值(3)不算。返回 (赢家, 三连下标元组) 或 (0, None)。"""
    for a, b, c in WIN_LINES:
        s = states[a]
        if s in (CIRCLE, CROSS) and s == states[b] == states[c]:
            return s, (a, b, c)
    return 0, None


class Game:
    """一盘超级井字棋。落子前用 legal_moves() 校验，非法落子一律返回 False。"""

    def __init__(self):
        self.cells = [[EMPTY] * 9 for _ in range(9)]
        self.grids = [GRID_OPEN] * 9
        self.turn = CIRCLE
        self.forced: int | None = None
        self.last_move: tuple[int, int] | None = None
        self.winner = 0
        self.win_line = None

    # ------------------------------------------------------------------ 查询
    def undecided_grids(self) -> list:
        return [i for i in range(9) if self.grids[i] == GRID_OPEN]

    def legal_moves(self) -> list:
        """当前全部合法落子 [(sub, cell), ...]。"""
        if self.winner:
            return []
        subs = self.undecided_grids()
        if self.forced is not None and self.grids[self.forced] == GRID_OPEN:
            subs = [self.forced]
        moves = []
        for s in subs:
            row = self.cells[s]
            for c in range(9):
                if row[c] == EMPTY:
                    moves.append((s, c))
        return moves

    def is_over(self) -> bool:
        return self.winner != 0

    # ------------------------------------------------------------------ 落子
    def apply_move(self, sub: int, cell: int) -> bool:
        """落子并推进局面。非法返回 False（局面不变）。"""
        if self.winner:
            return False
        if not (0 <= sub < 9 and 0 <= cell < 9):
            return False
        if self.cells[sub][cell] != EMPTY or self.grids[sub] != GRID_OPEN:
            return False
        if self.forced is not None and self.grids[self.forced] == GRID_OPEN \
                and sub != self.forced:
            return False

        self.cells[sub][cell] = self.turn
        self.last_move = (sub, cell)

        # 小格状态
        w, _ = line_winner(self.cells[sub])
        if w:
            self.grids[sub] = w
        elif all(self.cells[sub]):
            self.grids[sub] = GRID_TIE

        # 大棋盘胜负 / 平局
        w, line = line_winner(self.grids)
        if w:
            self.winner, self.win_line = w, line
        elif all(g != GRID_OPEN for g in self.grids):
            self.winner = GRID_TIE

        if not self.winner:
            # 下一手的强制格 = 本次落子的小格序号；该大格已决出则自由落子
            self.forced = cell if self.grids[cell] == GRID_OPEN else None
            self.turn = CROSS if self.turn == CIRCLE else CIRCLE
        return True

    # ------------------------------------------------------------------ 快照
    def snapshot(self) -> tuple:
        """不可变快照，供 AI 搜索使用。"""
        return (tuple(tuple(r) for r in self.cells), tuple(self.grids),
                self.forced, self.turn)

    def restore(self, snap: tuple) -> None:
        cells, grids, forced, turn = snap
        self.cells = [list(r) for r in cells]
        self.grids = list(grids)
        self.forced = forced
        self.turn = turn
        self.last_move = None
        self.winner = 0
        self.win_line = None
        w, line = line_winner(self.grids)
        if w:
            self.winner, self.win_line = w, line
        elif all(g != GRID_OPEN for g in self.grids):
            self.winner = GRID_TIE

File: rust/src/engine.rs
//! 超级井字棋（Ultimate Tic-Tac-Toe）规则引擎。
//! 语义与 Python 版 super_ttt/engine.py 逐条对齐（tests/test_engine.py 的
//! 全部用例已移植为 Rust 单元测试）。
//!
//! 值约定：格 0=空 1=圈 2=叉；大格 0=未决出 1=圈 2=叉 3=平。

#![allow(dead_code)]

pub const EMPTY: u8 = 0;
pub const CIRCLE: u8 = 1;
pub const CROSS: u8 = 2;
pub const GRID_OPEN: u8 = 0;
pub const GRID_TIE: u8 = 3;

pub const WIN_LINES: [[usize; 3]; 8] = [
    [0, 1, 2], [3, 4, 5], [6, 7, 8],
    [0, 3, 6], [1, 4, 7], [2, 5, 8],
    [0, 4, 8], [2, 4, 6],
];

/// 哨兵：线表不足 4 条时填充（合法相对下标 0..=8）
const SENT: u8 = 9;

const fn build_line_tables() -> ([[[u8; 3]; 4]; 9], [usize; 9]) {
    let mut out = [[[SENT, SENT, SENT]; 4]; 9];
    let mut counts = [2usize; 9];
    let mut c = 0;
    while c < 9 {
        let r = c / 3;
        let col = c % 3;
        out[c][0] = [(3 * r) as u8, (3 * r + 1) as u8, (3 * r + 2) as u8];
        out[c][1] = [col as u8, (col + 3) as u8, (col + 6) as u8];
        let mut n = 2;
        if r == col {
            out[c][n] = [0, 4, 8];
            n += 1;
        }
        if r + col == 2 {
            out[c][n] = [2, 4, 6];
            n += 1;
        }
        counts[c] = n;
        c += 1;
    }
    (out, counts)
}

/// CELL_LINES[c][li] = 包含小格 c 的线（相对下标，<=4 条，哨兵填充）
pub const CELL_LINES: [[[u8; 3]; 4]; 9] = build_line_tables().0;
pub const CELL_LINE_COUNT: [usize; 9] = build_line_tables().1;
/// 大棋盘与 3x3 几何相同，直接复用
pub const GRID_LINES: [[[u8; 3]; 4]; 9] = CELL_LINES;
pub const GRID_LINE_COUNT: [usize; 9] = CELL_LINE_COUNT;

/// 大棋盘胜负：返回 1/2（三连）、3（全部决出无胜者）、0（进行中）。
#[inline]
pub fn big_winner(grids: &[u8; 9]) -> u8 {
    for [a, b, c] in WIN_LINES {
        let s = grids[a];
        if s != 0 && s != GRID_TIE && s == grids[b] && s == grids[c] {
            return s;
        }
    }
    for g in grids {
        if *g == GRID_OPEN {
            return 0;
        }
    }
    GRID_TIE
}

/// 全量线检查（与 Python line_winner 等价）：返回 (赢家, 三连下标)。
pub fn line_winner(states: &[u8; 9]) -> (u8, Option<(u8, u8, u8)>) {
    for [a, b, c] in WIN_LINES {
        let s = states[a];
        if (s == CIRCLE || s == CROSS) && s == states[b] && s == states[c] {
            return (s, Some((a as u8, b as u8, c as u8)));
        }
    }
    (0, None)
}

#[derive(Clone, Copy)]
pub struct Game {
    pub cells: [[u8; 9]; 9],
    pub grids: [u8; 9],
    pub turn: u8,
    /// -1 = 自由落子（对应 Python None）
    pub forced: i8,
    pub last_move: Option<(u8, u8)>,
    pub winner: u8,
    pub win_line: Option<(u8, u8, u8)>,
}

impl Default for Game {
    fn default() -> Self {
        Self::new()
    }
}

impl Game {
    pub const fn new() -> Self {
        Game {
            cells: [[EMPTY; 9]; 9],
            grids: [GRID_OPEN; 9],
            turn: CIRCLE,
            forced: -1,
            last_move: None,
            winner: 0,
            win_line: None,
        }
    }

    pub fn is_over(&self) -> bool {
        self.winner != 0
    }

    /// 当前全部合法落子（sub-major 顺序，与 Python 版一致）。
    pub fn legal_moves(&self) -> Vec<(u8, u8)> {
        let mut moves = Vec::new();
        if self.winner != 0 {
            return moves;
        }
        let forced_sub: Option<usize> = if self.forced >= 0
            && self.grids[self.forced as usize] == GRID_OPEN
        {
            Some(self.forced as usize)
        } else {
            None
        };
        let subs: Vec<usize> = match forced_sub {
            Some(s) => vec![s],
            None => (0..9).filter(|&s| self.grids[s] == GRID_OPEN).collect(),
        };
        for s in subs {
            for c in 0..9 {
                if self.cells[s][c] == EMPTY {
                    moves.push((s as u8, c as u8));
                }
            }
        }
        moves
    }

    /// 落子并推进局面。非法返回 false（局面不变）。语义与 Python apply_move 对齐。
    pub fn apply_move(&mut self, sub: usize, cell: usize) -> bool {
        if self.winner != 0 {
            return false;
        }
        if sub >= 9 || cell >= 9 {
            return false;
        }
        if self.cells[sub][cell] != EMPTY || self.grids[sub] != GRID_OPEN {
            return false;
        }
        if self.forced >= 0
            && self.grids[self.forced as usize] == GRID_OPEN
            && sub as i8 != self.forced
        {
            return false;
        }

        self.cells[sub][cell] = self.turn;
        self.last_move = Some((sub as u8, cell as u8));

        // 小格状态
        let w = line_winner(&self.cells[sub]).0;
        if w != 0 {
            self.grids[sub] = w;
        } else if self.cells[sub].iter().all(|&v| v != EMPTY) {
            self.grids[sub] = GRID_TIE;
        }

        // 大棋盘胜负 / 平局
        let (w, line) = line_winner(&self.grids);
        if w != 0 {
            self.winner = w;
            self.win_line = line;
        } else if self.grids.iter().all(|&g| g != GRID_OPEN) {
            self.winner = GRID_TIE;
        }

        if self.winner == 0 {
            self.forced = if self.grids[cell] == GRID_OPEN {
                cell as i8
            } else {
                -1
            };
            self.turn = if self.turn == CIRCLE { CROSS } else { CIRCLE };
        }
        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn initial_81_legal_moves() {
        let g = Game::new();
        assert_eq!(g.legal_moves().len(), 81);
        assert_eq!(g.turn, CIRCLE);
        assert_eq!(g.forced, -1);
        assert!(!g.is_over());
    }

    #[test]
    fn forced_grid() {
        let mut g = Game::new();
        assert!(g.apply_move(0, 3));
        assert_eq!(g.forced, 3);
        assert_eq!(g.turn, CROSS);
        let moves = g.legal_moves();
        assert_eq!(moves.len(), 9);
        assert!(moves.iter().all(|m| m.0 == 3));
    }

    #[test]
    fn free_play_after_decided_target() {
        let mut g = Game::new();
        g.cells[3] = [CIRCLE; 9];
        g.grids[3] = CIRCLE;
        assert!(g.apply_move(0, 3));
        assert_eq!(g.forced, -1);
        assert_eq!(g.legal_moves().len(), 71);
    }

    #[test]
    fn illegal_moves() {
        let mut g = Game::new();
        assert!(g.apply_move(0, 0));
        assert!(!g.apply_move(1, 0)); // 强制格是 0
        assert_eq!(g.turn, CROSS);
        assert!(g.apply_move(0, 5));
        assert!(!g.apply_move(0, 5)); // 重复落子
        assert!(g.apply_move(5, 0));
        assert!(!g.apply_move(0, 5)); // 仍被占

        let mut g2 = Game::new();
        g2.cells[2] = [CIRCLE; 9];
        g2.grids[2] = CIRCLE;
        g2.forced = 2;
        assert!(!g2.apply_move(2, 0)); // 已决出大格不可落子
        assert!(g2.apply_move(0, 0));
    }

    #[test]
    fn big_board_win_and_tie() {
        let mut g = Game::new();
        for sub in [0usize, 4, 8] {
            g.cells[sub] = [CIRCLE; 9];
            g.grids[sub] = CIRCLE;
        }
        // 模拟 restore 重算
        let (w, line) = line_winner(&g.grids);
        g.winner = w;
        g.win_line = line;
        assert_eq!(g.winner, CIRCLE);
        assert_eq!(g.win_line, Some((0, 4, 8)));
        assert!(g.is_over());
        assert!(g.legal_moves().is_empty());

        let mut t = Game::new();
        t.grids = [GRID_TIE; 9];
        assert_eq!(big_winner(&t.grids), GRID_TIE);

        // 平局值 3 不构成三连
        let mut nt = Game::new();
        nt.grids[0] = GRID_TIE;
        nt.grids[1] = GRID_TIE;
        nt.grids[2] = GRID_TIE;
        assert_eq!(big_winner(&nt.grids), 0);
    }

    #[test]
    fn subgrid_full_without_winner_is_tie() {
        let mut g = Game::new();
        g.cells[5] = [0, CIRCLE, CROSS, CROSS, CROSS, CIRCLE, CIRCLE, CROSS, CIRCLE];
        assert!(g.apply_move(5, 0));
        assert_eq!(g.grids[5], GRID_TIE);
    }
}
